src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,264 @@
 #include "distributed/multi_utility.h"
 #include "distributed/worker_protocol.h"
 #include "executor/execdebug.h"
+#include "executor/executor.h"
+#include "commands/copy.h"
+#include "nodes/makefuncs.h"
 #include "storage/lmgr.h"
 #include "tcop/utility.h"
 #include "utils/snapmgr.h"
-
-
-static void CopyQueryResults(List *masterCopyStmtList);
+#include "utils/memutils.h"
 
 
 /*
- * multi_ExecutorStart is a hook called at at the beginning of any execution
- * of any query plan.
+ * FIXME: It'd probably be better to have different set of methods for:
+ * - router readonly queries
+ * - router modify
+ * - router insert ... select
+ * - real-time/task-tracker (no point in seperating those)
  *
- * If a distributed relation is the target of the query, perform some validity
- * checks. If a legal statement, start the distributed execution. After that
- * the to-be-executed query is replaced with the portion executing solely on
- * the master.
+ * I think it's better however to only have one type of CitusScanState, to
+ * allow to easily share code between routines.
  */
-void
-multi_ExecutorStart(QueryDesc *queryDesc, int eflags)
+static CustomExecMethods CitusCustomExecMethods = {
```

Comment:
I'd rather not have this named Citus vs. Router - that seems unclear. Router vs. RealTime?